### PR TITLE
[feat] implement routing to plant page

### LINKED ISSUE #35: Implement Routing to PlantPage
https://linear.app/trap-garden/issue/TG-35/implement-routing-to-plant-page

### PR DESCRIPTION
**Pull request by SashankBalusu**
### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- when user clicks on a plant they are routed to a new page. In that page they either see just the plant object if they were in all plants or the plant object and the user plant row from db if in my plants
- Refactored backend calls in view-plants to be solely present in the api folder
- Created backend functions to fetch userplants table by id and user id

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."
Screenshot 2024-11-11 at 6 01 19 PM

[image placeholder]: 

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
- can look through view-plants first and then the dynamic routing files under [plantId] and [userPlantId]
- changes to backend files are in userPlants and plants

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

### Online sources
[//]: # 'Copy links to any tutorials or documentation that was useful to you when working on this PR'

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @ccatherinetan

### CODE REVIEW

File: api/supabase/queries/plants.ts [hunk 1]
```diff
@@ -21,3 +21,8 @@ export async function getPlantById(plantId: UUID): Promise<Plant> {
 
   return data;
 }
+export async function getMatchingPlantForUserPlants(
```
Comment (ccatherinetan) [inline api/supabase/queries/plants.ts]: nit: maybe remove the plural at the end so `getMatchingPlantForUserPlant` 

honestly, we should rename `UserPlants` type to `UserPlant`

File: api/supabase/queries/userPlants.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    });
+    if (error) throw new Error(`Error inserting data: ${error.message}`);
+  });
+}
+export async function getUserPlantById(userPlantId: UUID): Promise<UserPlants> {
+  const { data, error } = await supabase
+    .from('user_plants')
+    .select('*')
+    .eq('id', userPlantId)
+    .is('date_harvested', null)
```
Comment (ccatherinetan) [inline api/supabase/queries/userPlants.ts]: we should rename this to `date_removed`; if i haven't changed it on supabase yet, go ahead and do that!

File: app/my-plants/[userPlantId]/page.tsx
Comment (ccatherinetan) [inline app/my-plants/[userPlantId]/page.tsx:1]: let's rename this route to `app/my-garden/[userPlantId]/page.tsx`

File: app/my-plants/[userPlantId]/page.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  const [currentPlant, setCurrentPlant] = useState<Plant>();
+  const [currentUserPlant, setCurrentUserPlant] = useState<UserPlants>();
+  useEffect(() => {
+    const getPlant = async () => {
+      const userPlant = await getUserPlantById(userPlantId);
+      setCurrentUserPlant(userPlant);
+      const plant = await getMatchingPlantForUserPlants(userPlant);
+      setCurrentPlant(plant);
+      // const plant = await getPlantById(await getUserPlantById(userPlantId))
+      // setCurrentPlant(plant)
```
Comment (ccatherinetan) [inline app/my-plants/[userPlantId]/page.tsx]: let's remove comments before merging!

File: app/view-plants/page.tsx [hunk 4]
```diff
@@ -1,41 +1,51 @@
...
   const [userPlants, setUserPlants] = useState<Plant[]>([]);
+  const [selectedPlants, setSelectedPlants] = useState<Plant[]>([]);
+  const [ownedPlants, setOwnedPlants] = useState<Map<UUID, UUID>>();
+
   const user_id: UUID = 'e72af66d-7aae-45f6-935a-187197749d9f';
   const userState = 'TENNESSEE';
+
+  const InlineDiv = styled.div`
+    display: inline-block;
+  `;
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: let's moving styling outside of the component. since we'll eventually have more styling for the view-plants page, let's move all styling into `app/view-plants/styles.ts`

see later comment: we may be able to remove this if we use onClick from PlantCard

File: api/supabase/queries/userPlants.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import { UUID } from 'crypto';
+import { UserPlants } from '@/types/schema';
+import supabase from '../createClient';
+
+export async function updateUserPlants(
```
Comment (ccatherinetan) [inline api/supabase/queries/userPlants.ts]: maybe rename this to insertUserPlants

File: api/supabase/queries/userPlants.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  return data;
+}
+export async function getUserPlantsByUserId(
+  user_id: UUID,
+): Promise<UserPlants[]> {
+  const { data, error } = await supabase
+    .from('user_plants')
+    .select('*')
+    .eq('user_id', user_id)
+    .is('date_harvested', null);
```
Comment (ccatherinetan) [inline api/supabase/queries/userPlants.ts]: this should be changed to 'date_removed' as well

File: app/view-plants/page.tsx [hunk 7]
```diff
@@ -1,41 +1,51 @@
...
-    const plantIds = data.map(row => row.plant_id) || [];
+    data.map(row => {
+      setOwnedPlants(prevOwnedPlants =>
+        new Map(prevOwnedPlants).set(row.plant_id, row.id),
+      );
+    });
 
     const plantsUser: Plant[] = await Promise.all(
-      plantIds.map(plantId => getPlantById(plantId)),
+      data.map(row => getMatchingPlantForUserPlants(row)),
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: here and on line 36, let's rename `row` to smth more descriptive. in this case you're mapping `userPlant`'s to `plant`'s, so row should be renamed to `userPlant`

File: app/view-plants/page.tsx [hunk 8]
```diff
@@ -1,41 +1,51 @@
...
+    display: inline-block;
+  `;
+
   async function fetchUserPlants(user_id: UUID) {
-    const { data, error } = await supabase
-      .from('user_plants')
-      .select('plant_id')
-      .eq('user_id', user_id)
-      .is('date_harvested', null);
+    const data = await getUserPlantsByUserId(user_id);
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: let's rename `data` to `fetchedUserPlants`

File: app/view-plants/page.tsx [hunk 9]
```diff
@@ -1,41 +1,51 @@
...
-      console.error('Error fetching plant IDs:', error);
-      return [];
-    }
-    if (!data) return [];
-    const plantIds = data.map(row => row.plant_id) || [];
+    data.map(row => {
+      setOwnedPlants(prevOwnedPlants =>
+        new Map(prevOwnedPlants).set(row.plant_id, row.id),
+      );
+    });
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: instead of updating the state each time and creating a new Map, consider instantiating the map, and creating the entire map object before calling `setOwnedPlants(idMap)` one time

however, i was thinking that we could create a new type OwnedPlant (below), since this would allow us to track the plantId for each userPlantId. Currently, if a user has duplicates of the same plant, the key/value will in the Map would be overwritten. While we prob won't support duplicate plant cards, it think the logic is cleaner if we define this new type instead of creating a map between UUIDs. 
```
interface OwnedPlant {
   userPlantId: UUID; 
   plant: Plant; 
}
```
If we follow my suggestion, we'd remove the userPlants state and refactor the ownedPlants state like so: 
`const [ownedPlants, setOwnedPlants] = useState<OwnedPlants[]>([]);`

File: app/view-plants/page.tsx [hunk 10]
```diff
@@ -1,41 +1,51 @@
...
     const plantsUser: Plant[] = await Promise.all(
-      plantIds.map(plantId => getPlantById(plantId)),
+      data.map(row => getMatchingPlantForUserPlants(row)),
     );
+
     return plantsUser;
   }
+
   useEffect(() => {
     const fetchPlantSeasonality = async () => {
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: we can rename this to `fetchPlants`. 
another way to simplify syntax is to directly call this function without naming it, i.e. 

```
(async () => {
      const plantList = await getAllPlants();
      const result = plantList.filter(plant => plant.us_state === userState);
      setPlants(result);
 })();
```
this can also be done for other funcs in useEffect's.

File: app/view-plants/page.tsx [hunk 11]
```diff
@@ -45,6 +55,7 @@ export default function Page() {
 
     fetchPlantSeasonality();
   }, []);
+
   useEffect(() => {
     const fetchData = async () => {
       const result = await fetchUserPlants(user_id);
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: if we follow my suggestion from earlier, we'd be calling `setOwnedPlants(result)` on line 62 instead

File: app/view-plants/page.tsx [hunk 12]
```diff
@@ -53,6 +64,28 @@ export default function Page() {
     fetchData();
   }, []);
 
+  function handlePlantCardClick(plant: Plant) {
+    //if in my plants go to [userPlantId] page
+    if (viewingOption == 'myPlants' && ownedPlants) {
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: === instead of ==

File: app/view-plants/page.tsx [hunk 13]
```diff
@@ -53,6 +64,28 @@ export default function Page() {
     fetchData();
   }, []);
 
+  function handlePlantCardClick(plant: Plant) {
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx:133]: if you follow from suggestion from earlier w the new OwnedPlant type, we can split up this function into
1. `handleUserPlantCardClick(ownedPlant: OwnedPlant)` to handle clicks on "My Garden" 
2. `handlePlantCardClick(plant: Plant)` to handle clicks on "All Plants"

File: app/view-plants/page.tsx [hunk 14]
```diff
@@ -67,7 +100,12 @@ export default function Page() {
             (userPlants.length ? (
               <div>
                 {userPlants.map((plant, key) => (
-                  <PlantCard key={key} plant={plant} canSelect={false} />
+                  <InlineDiv
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: is this only here to make `PlantCard` clickable? Why can't we directly use the `onClick` attribute of PlantCard?

File: components/PlantCard/styles.ts
Comment (ccatherinetan) [inline components/PlantCard/styles.ts:1]: see comments on Style PlantCard pr!

File: app/view-plants/page.tsx [hunk 15]
```diff
@@ -1,41 +1,51 @@
...
 export default function Page() {
+  const router = useRouter();
   const [viewingOption, setViewingOption] = useState<'myPlants' | 'all'>(
     'myPlants',
   );
   const [inAddMode, setInAddMode] = useState<boolean>(false);
-
   const [plants, setPlants] = useState<Plant[]>([]);
   const [userPlants, setUserPlants] = useState<Plant[]>([]);
+  const [selectedPlants, setSelectedPlants] = useState<Plant[]>([]);
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx:41]: note: unless we want to track the order in which someone selects plants, we could change this to a Set<Plant>, (see below), which might be slightly more efficient if there are a lot of user plants? 
` const [selectedPlants, setSelectedPlants] = useState<Set<Plant>>(new Set());`

I don't think we need to change it now, but just flagging it

File: app/view-plants/page.tsx [hunk 16]
```diff
@@ -80,8 +118,23 @@ export default function Page() {
           {viewingOption === 'all' &&
             (inAddMode ? (
               <div>
+                <h3>
+                  {selectedPlants.length > 0
+                    ? `${selectedPlants.length} Plants Selected`
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: nit: maybe we can rename the text to "Plant(s)" for grammar's sake - or we can make a util function that handles plurals

File: app/view-plants/page.tsx [hunk 17]
```diff
@@ -80,8 +118,23 @@ export default function Page() {
...
+                  >
+                    <PlantCard
+                      key={key}
+                      plant={plant}
+                      canSelect={true}
+                      isSelected={selectedPlants.includes(plant)}
+                    />
+                  </InlineDiv>
                 ))}
                 <div className="footer">
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: we should create style in `view-plants/styles.ts` called `Footer` rather than using className here
also, the select button should be at the top in the same row as "My Garden" and "All" (but that can be for the Style PlantCard PR or a future styling PR)

File: app/view-plants/page.tsx [hunk 18]
```diff
@@ -92,7 +145,12 @@ export default function Page() {
             ) : (
               <div>
                 {plants.map((plant, key) => (
-                  <PlantCard key={key} plant={plant} canSelect={false} />
+                  <InlineDiv
+                    key={key}
+                    onClick={() => handlePlantCardClick(plant)}
+                  >
+                    <PlantCard key={key} plant={plant} canSelect={false} />
+                  </InlineDiv>
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: note: we might be able to remove the outer ternary statement just use inAddMode where necessary, since the list of PlantCards is the same each time.

i.e. `canSelect={inAddMode}`,  etc.

but honestly, the current way might be more readable than splitting up the conditioning on inAddMode, so maybe we can leave it

File: components/PlantCard/index.tsx [hunk 19]
```diff
@@ -1,43 +1,60 @@
...
+  PlantAttributes,
+  RoundCheck,
+  TopRight,
+} from './styles';
 
 export default function PlantCard({
   plant,
   canSelect,
+  isSelected,
+  onToggleSelect,
```
Comment (ccatherinetan) [inline components/PlantCard/index.tsx]: 1. what if we changed `onToggleSelect` to `onClick`? or 
2. what if we added `onClick` as another prop? then on line 27, we could do `onClick={canSelect ? onToggleSelect : onClick}`

**Review by ccatherinetan — COMMENTED** [review]
looking great! the functionality is fantastic 🤩  A couple big changes to consider 
1. Create an OwnedPlant type and track a state of OwnedPlant[] instead of a map of UUIDs
2. split handlePlantCardClick into 2 functions that take in a OwnedPlant and a Plant to handle clicking on User PlantCards and PlantCards respectively 
3. remove InlineDiv and handle clicks all within PlantCard

File: components/PlantCard/index.tsx [hunk 20]
```diff
@@ -1,43 +1,56 @@
...
 
 export default function PlantCard({
   plant,
   canSelect,
+  isSelected,
+  onClick,
 }: {
   plant: Plant;
   canSelect: boolean;
+  isSelected?: boolean;
```
Comment (ccatherinetan) [inline components/PlantCard/index.tsx:21]: if we're making this an optional prop, perhaps we should pass a default value? 

```
export default function PlantCard({
  plant,
  canSelect,
  isSelected = false,
  onClick,
}: {
  plant: Plant;
  canSelect: boolean;
  isSelected?: boolean;
 onClick?: () => void;
});
```

File: api/supabase/queries/userPlants.ts [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    const { error } = await supabase.from('user_plants').insert({
+      user_id: userId,
+      plant_id: curr['plant_id'],
+      date_added: curr['date_added'],
+      date_removed: null,
+      planting_type: curr['planting_type'],
+    });
+    if (error) throw new Error(`Error inserting data: ${error.message}`);
+  });
+}
```
Comment (ccatherinetan) [inline api/supabase/queries/userPlants.ts:19]: Note: This comment requires refactoring the add-details flow, which is independent of this PR, so we don't have to implement it here. I'll prob make a separate PR to refactor add-details once this PR is merged. 

we should refactor this query to take in a `UserPlant[]` instead of a `Partial<UserPlant>[]`. Since our `UserPlant` should match the user_plants table schema on supabase, we can directly insert the entire UserPlant[] at once, instead of mapping it. This also makes this query more resilient to schema changes in the future, since the schema changes would be handled in the UserPlant type rather than inside the query. So, it should ultimately look like this: 

```
export async function insertUserPlants(
  userId: UUID,
  userPlants: UserPlant[],
) {
   const { data, error } = await supabase.from('user_plants').insert(userPlants).select();
   if (error) throw new Error(`Error inserting user plants: ${error.message}`);
   return data // an extra check! 
  });
}
```

see IJP's example [query](https://github.com/calblueprint/immigration-justice-project/blob/main/src/api/supabase/queries/profiles.ts)

File: api/supabase/queries/userPlants.ts [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    .is('date_removed', null)
+    .limit(1)
+    .single();
+
+  if (error) {
+    throw new Error(`Error fetching plant ID: ${error}`);
+  }
+  return data;
+}
+export async function getUserPlantsByUserId(
```
Comment (ccatherinetan) [inline api/supabase/queries/userPlants.ts]: nit: Maybe we should rename this to `getCurrentUserPlantsByUserId` since we're checking date_removed is not null -- just in case we want a `getAllUserPlantsByUserId` query in the future .

jk, we can leave it for now!

File: app/view-plants/page.tsx [hunk 23]
```diff
@@ -1,58 +1,77 @@
...
       const result = plantList.filter(plant => plant.us_state === userState);
       setPlants(result);
-    };
-
-    fetchPlantSeasonality();
+    })();
   }, []);
+
   useEffect(() => {
     const fetchData = async () => {
```
Comment (ccatherinetan) [inline app/view-plants/page.tsx]: can we directly just move fetchUserPlants into the useEffect instead of using fetchData as another wrapper?

so it could look like 
```
useEffect(() => {
  const fetchUserPlants = async () => {
    const fetchedUserPlants = await getUserPlantsByUserId(user_id);

    const ownedPlants: OwnedPlant[] = await Promise.all(
      fetchedUserPlants.map(async userPlant => {
        const plant = await getMatchingPlantForUserPlant(userPlant);
        return {
          userPlantId: userPlant.id,
          plant,
        };
      }),
    );

    setOwnedPlants(ownedPlants);
  }, []); // eventually when ProfileContext is connected, userId would be in the dependency list.
```

See [IJP](https://github.com/calblueprint/immigration-justice-project/blob/main/src/app/language-support/page.tsx) for an example of using queries in the useEffect

Comment (ccatherinetan) [reply app/view-plants/page.tsx]: note: since the queries (both `getUserPlantsByUserId` and `getMatchingPlantForUserPlant`) throw an error on error, we may want to use a try / catch block here and wherever a query is called to handle the error, i.e. console.error() it in the future (see IJP's example above)

**Review by ccatherinetan — COMMENTED** [review]
looking so great!! some final things but otherwise ready to merge!

File: api/supabase/queries/userPlants.ts (hunk 21 above)
Comment (SashankBalusu) [reply api/supabase/queries/userPlants.ts:19]: sounds good!
